state/unit.go

```diff
@@ -805,8 +817,23 @@ func (u *Unit) destroyHostOps(a *Application, force bool) (ops []txn.Op, errs []
 	machineUpdate := bson.D{{"$pull", bson.D{{"principals", u.doc.Name}}}}
 	var cleanupOps []txn.Op
 	if machineCheck && containerCheck {
-		machineUpdate = append(machineUpdate, bson.D{{"$set", bson.D{{"life", Dying}}}}...)
-		cleanupOps = []txn.Op{newCleanupOp(cleanupDyingMachine, m.doc.Id, force)}
+		err = m.advanceLifecycle(Dying, force)
```

Comment:
It's not valid to replace the ops with a call to advanceLifecycle here. The point of this code is to build a larger transaction that only advances the lifecycle if the rest of the transaction succeeds - with your change some other assertions could fail so the cleanup wouldn't be created, but the machine would still be set to dying. It makes it not an atomic transaction.

Reply:
Ok.
There are a few other places where the effect is immediate but I am happy to adjust this to accumulate operations as before in order to land this PR. 
I do feel that what we want to have a call to advance lifecycle here as there is a bit more of nice logic performed as part of this call and we'd want consistency of operations but I am happy to come back to it later as a separate work.